File: 동적 계획법/2156/2156.py
n = int(input())
arr, answer = [0] * (n + 3), [0] * (n + 3)
for i in range(3, n + 3):
   arr[i] = int(input())

for i in range(3, n + 3):
   answer[i] = max(answer[i - 1], arr[i] + max(answer[i - 3] + arr[i - 1], answer[i - 2]))
print(answer[n + 2])

File: 동적 계획법/2156/2156.cpp
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

int main()
{
   int n;
   cin>>n;
   vector<int> v(n + 3, 0), answer(n + 3, 0);
   for (int i = 3; i < n + 3; ++i)
      cin>>v[i];
   
   for (int i = 3; i < n + 3; ++i)
      answer[i] = max(answer[i - 1], v[i] + max(answer[i - 3] + v[i - 1], answer[i - 2]));
   cout<<answer[n + 2];

   return 0;
}